pysteps/blending/steps.py

```diff
@@ -560,336 +570,659 @@ def forecast(
+        # only nans or infs. If so, fill it with the zero value.
+        if zero_precip_radar:
+            precip_cascade = np.nan_to_num(
+                precip_cascade,
+                copy=True,
+                nan=np.nanmin(precip_models_cascade),
+                posinf=np.nanmin(precip_models_cascade),
+                neginf=np.nanmin(precip_models_cascade),
+            )
 
-    # 7. initizalize the current and previous extrapolation forecast scale
-    # for the nowcasting component
-    rho_extr_prev = np.repeat(1.0, PHI.shape[0])
-    rho_extr = PHI[:, 0] / (1.0 - PHI[:, 1])  # phi1 / (1 - phi2), see BPS2004
+        # 2.3.4 Check if the NWP fields contain nans or infinite numbers. If so,
+        # fill these with the minimum value present in precip (corresponding to
+        # zero rainfall in the radar observations)
+        (
+            precip_models_cascade,
+            precip_models_pm,
+            mu_models,
+            sigma_models,
+        ) = _fill_nans_infs_nwp_cascade(
+            precip_models_cascade,
+            precip_models_pm,
+            precip_cascade,
+            precip,
+            mu_models,
+            sigma_models,
+        )
 
-    if measure_time:
-        init_time = time.time() - starttime_init
+        # 2.3.5 If zero_precip_radar is True, only use the velocity field of the NWP
+        # forecast. I.e., velocity (radar) equals velocity_model at the first time
+        # step.
+        if zero_precip_radar:
+            # Use the velocity from velocity_models at time step 0
+            velocity = velocity_models[:, 0, :, :, :]
+            # Take the average over the first axis, which corresponds to n_models
+            # (hence, the model average)
+            velocity = np.mean(velocity, axis=0)
+
+        # 3. Initialize the noise method.
+        # If zero_precip_radar is True, initialize noise based on the NWP field time
+        # step where the fraction of rainy cells is highest (because other lead times
+        # might be zero as well). Else, initialize the noise with the radar
+        # rainfall data
+        if zero_precip_radar:
+            precip_noise_input = _determine_max_nr_rainy_cells_nwp(
+                precip_models_pm, precip_thr, precip_models_pm.shape[0], timesteps
+            )
+            # Make sure precip_noise_input is three dimensional
+            precip_noise_input = precip_noise_input[np.newaxis, :, :]
+        else:
+            precip_noise_input = precip.copy()
 
-    ###
-    # 8. Start the forecasting loop
-    ###
-    print("Starting blended nowcast computation.")
+        pp, generate_noise, noise_std_coeffs = _init_noise(
+            precip_noise_input,
+            precip_thr,
+            n_cascade_levels,
+            bp_filter,
+            decompositor,
+            fft,
+            noise_method,
+            noise_kwargs,
+            noise_stddev_adj,
+            measure_time,
+            num_workers,
+            seed,
```

Comment:
@mpvginde I added the missing `seed` parameter here, do we also need to set the seed with numpy before the call to `_init_noise` as done elsewhere in this file?